'socket hang up' error thrown randomly

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**- Do you want to request a *feature* or report a *bug*?**

Bug

**- What is the current behavior?**

At points during development the following error occurs which stops the entire Netlify Dev stack:

```
.../node_modules/netlify-redirector/lib/redirects.js:116
      throw ex;
      ^

Error: socket hang up
    at createHangUpError (_http_client.js:323:15)
    at Socket.socketCloseListener (_http_client.js:364:25)
    at Socket.emit (events.js:194:15)
    at TCP._handle.close (net.js:600:12)
```

Looks like it was also reported on the community forum back in April - https://community.netlify.com/t/netlify-dev-regularly-crashes/551

**- If the current behavior is a bug, please provide the steps to reproduce.**

I haven't been able see any common pattern for what causes this. 

**- What is the expected behavior?**

Should not crash. 

**- Please mention your node.js, and operating system version.**

- Node: `v10.15.1`
- Using `yarn`
- Netlify CLI `2.12.0`

Running into this issue since forever, too. It's not consistent at all. For all I know, it happens when I save a change which triggers a reload, and then do something during the reload that causes the error to throw. Might be completely wrong though, it's just my hypothesis that appears to work, i.e. avoid seeing this error more often. 😆 

```
/Users/mrlubos/.nvm/versions/node/v12.14.1/lib/node_modules/netlify-cli/node_modules/netlify-redirector/lib/redirects.js:116
      throw ex;
      ^

Error: socket hang up
    at connResetException (internal/errors.js:570:14)
    at Socket.socketCloseListener (_http_client.js:380:25)
    at Socket.emit (events.js:228:7)
    at Socket.EventEmitter.emit (domain.js:475:20)
    at TCP.<anonymous> (net.js:664:12) {
  code: 'ECONNRESET'
}
npm ERR! code ELIFECYCLE
npm ERR! errno 7
```

--- a/src/commands/dev/index.js
+++ b/src/commands/dev/index.js
@@ -5,6 +5,8 @@ const process = require('process')
 const { flags: flagsLib } = require('@oclif/command')
 const boxen = require('boxen')
 const chalk = require('chalk')
+const fetch = require('node-fetch')
+const waitFor = require('p-wait-for')
 const StaticServer = require('static-server')
 const stripAnsiCc = require('strip-ansi-control-characters')
 const waitPort = require('wait-port')
@@ -20,6 +22,13 @@ const openBrowser = require('../../utils/open-browser')
 const { startProxy } = require('../../utils/proxy')
 const { startFunctionsServer } = require('../../utils/serve-functions')
 const { startForwardProxy } = require('../../utils/traffic-mesh')
+
+// 1 second
+const SERVER_POLL_INTERVAL = 1e3
+// 150 milliseconds
+const SERVER_POLL_REQUEST_TIMEOUT = 150
+// 30 seconds
+const SERVER_POLL_TIMEOUT = 3e4
 
 const startFrameworkServer = async function ({ settings, log, exit }) {
   if (settings.noCmd) {
@@ -82,7 +91,30 @@ const startFrameworkServer = async function ({ settings, log, exit }) {
   })
 
   try {
-    const open = await waitPort({ port: settings.frameworkPort, output: 'silent', timeout: FRAMEWORK_PORT_TIMEOUT })
+    const open = await waitPort({
+      port: settings.frameworkPort,
+      output: 'silent',
+      timeout: FRAMEWORK_PORT_TIMEOUT,
+    })
+
+    const waitForServerToRespond = async () => {
+      try {
+        await fetch(`http://localhost:${settings.frameworkPort}`, {
+          method: 'HEAD',
+          timeout: SERVER_POLL_REQUEST_TIMEOUT,
+        })
+      } catch (_) {
+        return false
+      }
+
+      return true
+    }
+
+    await waitFor(waitForServerToRespond, {
+      interval: SERVER_POLL_INTERVAL,
+      timeout: SERVER_POLL_TIMEOUT,
+    })
+
     if (!open) {
       throw new Error(`Timed out waiting for port '${settings.frameworkPort}' to be open`)
     }

--- a/src/commands/dev/index.js
+++ b/src/commands/dev/index.js
@@ -27,8 +27,8 @@ const { startForwardProxy } = require('../../utils/traffic-mesh')
 const SERVER_POLL_INTERVAL = 1e3
 // 150 milliseconds
 const SERVER_POLL_REQUEST_TIMEOUT = 150
-// 30 seconds
-const SERVER_POLL_TIMEOUT = 3e4
+// 20 seconds
+const SERVER_POLL_TIMEOUT = 2e4
 
 const startFrameworkServer = async function ({ settings, log, exit }) {
   if (settings.noCmd) {
@@ -97,6 +97,10 @@ const startFrameworkServer = async function ({ settings, log, exit }) {
       timeout: FRAMEWORK_PORT_TIMEOUT,
     })
 
+    if (!open) {
+      throw new Error(`Timed out waiting for port '${settings.frameworkPort}' to be open`)
+    }
+
     const waitForServerToRespond = async () => {
       try {
         await fetch(`http://localhost:${settings.frameworkPort}`, {
@@ -110,13 +114,13 @@ const startFrameworkServer = async function ({ settings, log, exit }) {
       return true
     }
 
-    await waitFor(waitForServerToRespond, {
-      interval: SERVER_POLL_INTERVAL,
-      timeout: SERVER_POLL_TIMEOUT,
-    })
-
-    if (!open) {
-      throw new Error(`Timed out waiting for port '${settings.frameworkPort}' to be open`)
+    try {
+      await waitFor(waitForServerToRespond, {
+        interval: SERVER_POLL_INTERVAL,
+        timeout: SERVER_POLL_TIMEOUT,
+      })
+    } catch (_) {
+      log(NETLIFYDEVWARN, 'Netlify Dev could not verify that your framework server is responding to requests.')
     }
   } catch (error) {
     log(NETLIFYDEVERR, `Netlify Dev could not connect to localhost:${settings.frameworkPort}.`)

